File: Collections/ArrayList.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Collections
{
    /// <summary>
    /// List implementation using a dynamically-resizing array.
    /// </summary>
    public class ArrayList<T> : IList<T> where T : class
    {
        private T[] stack;
        private int head;

        public ArrayList(int capacity = 10)
        {
            if (capacity < 1)
            {
                throw new ArgumentOutOfRangeException();
            }
            stack = new T[capacity];
            head = -1;  // initially empty
        }

        public ArrayList(params T[] elements) : this(Math.Max(1, elements.Length))
        {
            foreach (var element in elements)
                Push(element);
        }

        public T this[int i]
        {
            get
            {
                if (i > Size())
                {
                    throw new ArgumentOutOfRangeException();
                }
                return stack[i];
            }

            set
            {
                if (i > Size())
                {
                    throw new ArgumentOutOfRangeException();
                }
                if (value == null)
                {
                    throw new ArgumentNullException();
                }
                stack[i] = value;
            }
        }

        public bool IsEmpty() => head == -1;

        private void Resize(int size)
        {
            var temp = new T[size];
            for (int i = 0; i <= head; i++)
            {
                temp[i] = stack[i];
            }
            stack = temp;
        }

        public T Peek()
        {
            if (IsEmpty())
            {
                throw new InvalidOperationException("Empty list");
            }
            return stack[head];
        }

        public T Pop()
        {
            if (IsEmpty())
            {
                throw new InvalidOperationException("Empty list");
            }

            // We actively de-reference to prevent loitering,
            // which is why the implementation can only support reference types.
            var returnValue = stack[head];
            stack[head--] = null;  // cannot be confused for a null payload as we always track the head.

            // Lazily reduce the size of the backing array if necessary
            if (head == stack.Length / 4)
            {
                Resize(stack.Length / 2);
            }

            return returnValue;
        }

        public void Push(T payload)
        {
            // Lazily increase the size of the backing array if necessary
            if (head == stack.Length - 1)
            {
                Resize(2 * stack.Length);
            }
            stack[++head] = payload;
        }

        public int Size() => head + 1;

        public int Capacity() => stack.Length;

        public IEnumerator<T> GetEnumerator()
        {
            for (int i = head; i > -1; i--)
            {
                yield return stack[i];
            }
        }

        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    }
}

File: src/main/java/in/bfdes/collections/ArrayList.java
package in.bfdes.collections;

import java.util.Iterator;
import java.util.NoSuchElementException;

public class ArrayList<T> implements List<T> {
    private T[] stack;
    private int head = -1;

    public ArrayList(int initialCapacity) {
        if (initialCapacity < 1)
            throw new IllegalArgumentException();
        stack = (T[]) new Object[initialCapacity];
    }

    public ArrayList() {
        this(10);
    }

    public ArrayList(T[] initialItems) {
        this(Math.max(1, initialItems.length));
        for (var item : initialItems)
            push(item);
    }

    public ArrayList(Iterable<T> initialItems) {
        for (var item : initialItems)
            push(item);
    }

    @Override
    public int size() {
        return head + 1;
    }

    @Override
    public boolean isEmpty() {
        return head == -1;
    }

    private boolean isFull() {
        return head == stack.length - 1;
    }

    @Override
    public T peek() {
        if (isEmpty())
            throw new NoSuchElementException();
        return stack[head];
    }

    @Override
    public T pop() {
        if (isEmpty())
            throw new NoSuchElementException();
        var item = stack[head];
        stack[head--] = null;
        if (head == stack.length / 4)
            resize(stack.length / 2);
        return item;
    }

    @Override
    public void push(T item) {
        if (item == null)
            throw new IllegalArgumentException();
        if (isFull())
            resize(2 * stack.length);
        stack[++head] = item;
    }

    private void resize(int newCapacity) {
        var newStack = (T[]) new Object[newCapacity];
        for (int i = 0; i <= head; i++)
            newStack[i] = stack[i];
        stack = newStack;
    }

    @Override
    public Iterator<T> iterator() {
        return new Iterator<>() {
            private int next = 0;

            @Override
            public boolean hasNext() {
                return next <= head;
            }

            @Override
            public T next() {
                if (!hasNext())
                    throw new NoSuchElementException();
                return stack[next++];
            }
        };
    }
}
